### PR TITLE
Refactor the architecture to actor based impl.

### LINKED ISSUE #78: Discard stale watch events
If compilation takes a lot of time, all recompile events should be discarded other than process them in order. Or to say, only the latest recompile request should be processed. Reported by aleph on discord.

### PR DESCRIPTION
**Pull request by Enter-tainer**
This PR should fix #78 #81 #79

TODO:
- [x] ~~when doc is empty, it will not respond to frontend's `current` request~~ This is not a problem
- [x] currently partial rendering doesn't work

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -144,51 +113,29 @@ impl SrcToDocJumpRequest {
...
+        // TODO: set this back to Info
+        .filter_module("typst_preview", log::LevelFilter::Debug)
         .filter_module("typst_ts", log::LevelFilter::Info)
         .try_init();
     let arguments = CliArguments::parse();
     info!("Arguments: {:#?}", arguments);
-    let doc_publisher: Publisher<Document> = PublisherImpl::new().into();
     let command = CompileSettings::with_arguments(arguments.clone());
-    let enable_partial_rendering = arguments.enable_partial_rendering;
+    let _enable_partial_rendering = arguments.enable_partial_rendering;
```
Comment (Enter-tainer) [inline src/main.rs:138]: should be re-enabled very soon

File: src/main.rs [hunk 2]
```diff
@@ -256,231 +209,58 @@ async fn main() {
...
+                tokio::spawn(async move {
+                    webview_actor.run().await;
+                });
+                let render_actor =
+                    actor::render::RenderActor::new(render_full.1, doc_watch_rx, svg.0);
+                std::thread::spawn(move || {
+                    render_actor.run();
+                });
+                let mut renderer_rx = renderer_mailbox.0.subscribe();
+                tokio::spawn(async move {
```
Comment (Enter-tainer) [inline src/main.rs:236]: it is not possible for a non-async thread to listen two tokio channels at once. (listen on 1 channel is possible, by using blocking_recv.) Given that render_actor has two message source, I spawn a tokio task to as a "router" to combine and foward message for it.

File: src/actor/editor.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                        ControlPlaneMessage::RemoveMemoryFiles(memory_files) => {
+                            debug!("EditorActor: received message from editor: RemoveMemoryFiles {:?}", &memory_files.files);
+                            self.world_sender.send(WorldActorRequest::RemoveMemoryFiles(memory_files))
+                        }
+                    }.unwrap();
+                }
+            }
+        }
+        info!("EditorActor: ws disconnected, shutting down whole program");
+        std::process::exit(0);
```
Comment (Enter-tainer) [inline src/actor/editor.rs:109]: when editor disconnects, it shutdown itself. this should effectively fix #81 . Although works in vscode extension code are left untouched now

File: src/main.rs (hunk 2 above)
Comment (Myriad-Dreamin) [reply src/main.rs:243]: Intuitively, it is not the bottleneck in rust. Also, `comemo` requires to be binded to a specific thread, as discussed.

**Review by Myriad-Dreamin — COMMENTED** [review]
reviewed

**Pull request merged**